### PR TITLE
feat(auth): support sgID for form submissions

### PR DESCRIPTION
**Pull request by LoneRifle**
## Problem
FormSG is to acquire an additional authentication means for form submitters, ie, sgID. This PR will provide these integrations

## Solution
Fully flesh out sgID integrations with FormSG

[config]
- Add needed configuration to configure sgID authentication, driving
  most of these using env vars

[modules]
- add an sgid module, taking inspiration from spcp
- add sgid-related entries alongside spcp equivalents in public-form
  and email-submission controllers
- do likewise for frontend components

**Breaking Changes** 
No - this PR is backwards compatible  

**Logistics**
-  This PR has been blocked by #1987; once that PR is merged to develop, this PR must be rebased on the branch

## Screenshots

https://user-images.githubusercontent.com/10572368/119929313-d80f5e80-bfaf-11eb-99c0-459b92baf9cf.mov

## Tests
- Find an sgID developer who has access to a special build of the Singpass App that talks to sgID staging
- Create a betaFlag - `sgid` - for a user
- With this user, create a form with auth type set to Singpass (Lite)
- Make the form public and attempt to log into the form

## Deploy Notes

**New environment variables**:
  - `SGID_ENDPOINT` - The sgID OIDC endpoint
  - `SGID_CLIENT_ID` - The client id registered with sgID
  - `SGID_CLIENT_SECRET` - The client secret registered with sgID
  - `SGID_PRIVATE_KEY` - The path to the private key to decrypt payloads from sgID. Also used for JWT signing
  - `SGID_PUBLIC_KEY` - The path to the public key given to sgID. Also used to verify JWTs created by FormSG
  - `SGID_REDIRECT_URI` - The callback uri that sgID will pass the authorization code and state to
  - `SGID_COOKIE_MAX_AGE` - Max sgID cookie age with remember me unchecked
  - `SGID_COOKIE_MAX_AGE_PRESERVED` - Max sgID cookie age with remember me checked
  - `SGID_COOKIE_DOMAIN` - Domain name set on cookie that holds the sgID jwt

**New dependencies**:

- `@opengovsg/sgid-client` : SDK for sgID based on openid-client

### CODE REVIEW

**Review by seaerchin — COMMENTED** [review]
cool stuff! cos this is a draft pr, not going to focus on code style etc but just wanted to check if my understanding is correct!

as i understand it, we hit the `submit-form-directive` and what happens is that this issues a GET request over to the backend. This isn't rendered because of the `ng-if` (except the `start-page`) component.

next, when we click login, there is a request issued to `/auth/redirect` to get the redirect url and subsequently, redirect there. Next, the sgID client sets the cookie on the requests and routes back. This then triggers a `setUser` call to set the `SpcpSession.userName`, which allows the 
form to be rendered. 

On submission, we verify and check for existence of a username field then attach it to the email response.

this should be roughly what the PR implements right?

**Comment by LoneRifle** [comment]
> cool stuff! cos this is a draft pr, not going to focus on code style etc but just wanted to check if my understanding is correct!
> 
> as i understand it, we hit the `submit-form-directive` and what happens is that this issues a GET request over to the backend. This isn't rendered because of the `ng-if` (except the `start-page`) component.
> 
> next, when we click login, there is a request issued to `/auth/redirect` to get the redirect url and subsequently, redirect there. Next, the sgID client sets the cookie on the requests and routes back. This then triggers a `setUser` call to set the `SpcpSession.userName`, which allows the
> form to be rendered.
> 
> On submission, we verify and check for existence of a username field then attach it to the email response.
> 
> this should be roughly what the PR implements right?

Indeed. This was basically almost a drop-in replacement for Singpass SAML, and a lot of code was re-purposed from the SAML integration.

File: src/app/modules/sgid/sgid.errors.ts
Comment (liangyuanruo) [inline src/app/modules/sgid/sgid.errors.ts:3]: suggest prefixing names with `Sgid` or similar for ease of interpretation against other errors in the logs

File: docker-compose.yml
Comment (liangyuanruo) [inline docker-compose.yml:78]: heads up - I _think_ we are running against the maximum length of 4096 characters for environment variables in Elastic Beanstalk. have you tried deploying to staging?

Comment (LoneRifle) [reply docker-compose.yml:78]: we haven't, given that there's still a bunch of setup to undertake. Thanks for notifying!

Comment (LoneRifle) [reply docker-compose.yml:78]: ... do we have a viable solution to the extensive configuration that FormSG needs? I wouldn't want to compete with the [Java runtime for number of env vars supported](http://bash.org/?14351)

**Comment by mantariksh** [comment]
Blocked by #1987

File: src/app/config/schema.ts
Comment (mantariksh) [inline src/app/config/schema.ts:244]: might it be better to put these in a separate file like in `spcp-myinfo.config`? might be easier to keep the slices of config more manageable

**Review by mantariksh — COMMENTED** [review]
I noticed the implementation depends on the frontend being able to decode the JWT and display the NRIC. we're actually planning on removing that behaviour in the frontend, since we return the NRIC via the `spcpSession` key from the backend anyway. this way we can increase the security of the cookie by setting `httpOnly` to true, since the frontend javascript has no need to read it. would this work for SGID?

File: src/app/config/schema.ts
Comment (LoneRifle) [reply src/app/config/schema.ts:244]: I don't mind, let me rework this

**Comment by LoneRifle** [comment]
> we're actually planning on removing that behaviour in the frontend, since we return the NRIC via the `spcpSession` key from the backend anyway. this way we can increase the security of the cookie by setting `httpOnly` to true, since the frontend javascript has no need to read it. would this work for SGID?

should be okay, this module is patterned after spcp anyway. should I wait for that feature introduction too?

File: src/app/config/config.ts
Comment (karrui) [inline src/app/config/config.ts:39]: nit: prefer if this instantiation resides in `app/config/features`, following the other features

File: src/app/config/config.ts
Comment (karrui) [inline src/app/config/config.ts:216]: if we move the `sgid` instantiation to its own file, then we will not need to add it into the exported `config` constant and `sgid.service` can import directly

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:11]: should the format be `url` instead? should we also set the default to `null` for clarity

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:17]: should the format be `String` with a `null` default?

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:23]: should the format be `String` with a `null` default?

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:29]: should the format be `String` with a `null` default?

should we also rename this to `privateKeyPath` for clarity?

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:35]: should the format be `String` with a `null` default?

should we also rename this to `publicKeyPath` for clarity?

File: src/app/config/sgid.ts
Comment (karrui) [inline src/app/config/sgid.ts:41]: should the format be `"url"` with a `null` default?

File: src/app/loaders/express/index.ts
Comment (karrui) [inline src/app/loaders/express/index.ts:164]: the singpass and corppass routes are stuck as they are due to a hardcoded NDI whitelist.

i don't think this is the case for `/sgid` right? Should we move it to a more semantically correct API endpoint `/api/v3/forms/:formId/auth/sgid` instead? 

the new router can reside in `PublicFormAuthRouter` :P

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:58]: can we make it more explicit on what `state` is? From the call in the controller it seems to be made up of `formId` and `isPersistentLogin`; should we make those arguments of this function and create the state string in this function instead?

There will also be a benefit where the controller does not need to know how to create a state and leave that to this function.

File: src/app/modules/sgid/__tests__/sgid.controller.spec.ts
Comment (karrui) [inline src/app/modules/sgid/__tests__/sgid.controller.spec.ts:35]: nit: PascalCase the service name (`SgidService`)?

File: src/app/modules/sgid/__tests__/sgid.controller.spec.ts
Comment (karrui) [inline src/app/modules/sgid/__tests__/sgid.controller.spec.ts:52]: nit: should we invoke the full `SgidController.handleLogin` instead of destructuring it here? Allows the reader to know where the function originates amongst all the other functions being called :P

File: src/app/modules/sgid/__tests__/sgid.routes.spec.ts
Comment (karrui) [inline src/app/modules/sgid/__tests__/sgid.routes.spec.ts:143]: wrong comment? looks like an SP form instead :P

File: src/app/modules/sgid/__tests__/sgid.service.spec.ts
Comment (karrui) [inline src/app/modules/sgid/__tests__/sgid.service.spec.ts:74]: 🤔

File: src/app/modules/sgid/sgid.controller.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.controller.ts:1]: we've removed `express-serve-static-core` from the app, use `ControllerHandler` from `'core/core.types'` instead? :P

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [reply src/app/modules/sgid/sgid.service.ts:58]: can we also have JSDocs for all the exported functions in this file

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:86]: nit: increase the clarity of what this function does with a more explicit name? Should it be called `retrieveAccessToken` instead?

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:105]: same ^, should we call this `retrieveUserInfo`?

File: src/app/modules/sgid/sgid.controller.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.controller.ts:76]: as noted by @mantariksh, @tshuli is working on #2193 which may necessitate some changes on this PR :P FYI

Comment (karrui) [reply src/app/modules/sgid/sgid.controller.ts:76]: OR we can ignore it for now and update it as part of #2193 instead. See who is faster :P

File: src/app/modules/sgid/sgid.util.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.util.ts:28]: the typeguard comment seems to insinuate that this checks for the existence of an eserviceId too. Should that be part of the checks in this typeguard?

EDIT: just saw that SGID doesn't have an esrvcId. Update the comment then?

File: src/app/modules/verified-content/verified-content.service.ts
Comment (karrui) [inline src/app/modules/verified-content/verified-content.service.ts:39]: iffy on this returning an error. Should we just constrain the type to be back to the previous implementation; i.e. only allow SP/CP auth types? 

Only add in here when SGID is ready to support verified fields?

EDIT: I see that there is a new `SgidUinFin` key being added? Does that not mean SGID has verified content now?

File: src/public/modules/forms/admin/directives/settings-form.client.directive.js
Comment (karrui) [inline src/public/modules/forms/admin/directives/settings-form.client.directive.js:5]: welcome back ol' BetaService 😆

**Review by karrui — CHANGES_REQUESTED** [review]
sorry for the deluge of comments

File: src/app/modules/verified-content/verified-content.service.ts
Comment (LoneRifle) [reply src/app/modules/verified-content/verified-content.service.ts:39]: It does... but only for e-mail mode. @liangyuanruo has disallowed sgID for providing verified content in storage mode until we have e2e encryption in-place for fields obtained from eg MyInfo and sgID

File: src/app/modules/sgid/__tests__/sgid.controller.spec.ts
Comment (LoneRifle) [reply src/app/modules/sgid/__tests__/sgid.controller.spec.ts:35]: It isn't a class definition though, but okay.

File: src/app/loaders/express/index.ts
Comment (LoneRifle) [reply src/app/loaders/express/index.ts:164]: The route was decided by @liangyuanruo . I have already raised the issue of placing the endpoint at the appropriate path.

File: src/app/modules/sgid/__tests__/sgid.service.spec.ts
Comment (LoneRifle) [reply src/app/modules/sgid/__tests__/sgid.service.spec.ts:74]: I need this there to induce SgidService to throw up

File: src/app/config/sgid.ts
Comment (LoneRifle) [reply src/app/config/sgid.ts]: I give up:
```
    'endpoint' uses an unknown format type: url
```

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:60]: nit: don't need types in JSDoc anymore, the Typescript typings should be the source of truth?

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:166]: nit: change name to `createJwt` instead to follow the current (unspoken and loosely enforced 😢 ) naming conventions in this repo?

File: src/app/modules/sgid/sgid.service.ts
Comment (karrui) [inline src/app/modules/sgid/sgid.service.ts:183]: nit: here too, change name to `createJwtInfo` instead?

**Review by karrui — APPROVED** [review]
lgtm with nits

File: src/app/config/config.ts
Comment (mantariksh) [inline src/app/config/config.ts:17]: we don't need to export this from the central config right? it can just be imported directly from the SGID config file when needed

File: src/public/modules/forms/admin/directives/settings-form.client.directive.js
Comment (mantariksh) [inline src/public/modules/forms/admin/directives/settings-form.client.directive.js:151]: this implementation looks like if the beta flag is not enabled, the user can see the SGID option but it is disabled for them right? might this be confusing for users who see a new option but don't know why it's disabled for them?

Comment (LoneRifle) [reply src/public/modules/forms/admin/directives/settings-form.client.directive.js:151]: It's possible, but this may be deliberate; if they send in a support query about it, FormSG can then direct that query to sgID.

File: src/app/config/config.ts
Comment (LoneRifle) [reply src/app/config/config.ts:17]: This is done to be consistent with the other feature config

File: src/app/modules/sgid/sgid.service.ts
Comment (LoneRifle) [reply src/app/modules/sgid/sgid.service.ts:183]: We're not creating anything here; we are extracting information out of a JWT

File: src/app/config/config.ts
Comment (mantariksh) [reply src/app/config/config.ts:17]: no leh, none of the other feature config is exported both from the config file and from the main config. e.g. `spcp-myinfo.config` is not.

**Comment by mantariksh** [comment]
awaiting confirmation on design and copy

**Comment by LoneRifle** [comment]
> awaiting confirmation on design and copy

Confirmation sought. .. I have noticed however that #2193 will be merged soon, and this branch will have to be rebased (again) to bring in-line with norms.

**Pull request merged**
